tensorboard/webapp/plugins/text/text_v2_plugin.py

```diff
@@ -0,0 +1,209 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Angular Version of Text Plugin."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import textwrap
+
+# Necessary for an internal test with special behavior for numpy.
+import numpy as np
+
+import six
+from werkzeug import wrappers
+
+from tensorboard import plugin_util
+from tensorboard.util import tensor_util
+from tensorboard.backend import http_util
+from tensorboard.plugins import base_plugin
+from tensorboard.data import provider
+from tensorboard.plugins.text import metadata
+
+# HTTP routes
+TAGS_ROUTE = "/tags"
+TEXT_ROUTE = "/text"
+
+_DEFAULT_DOWNSAMPLING = 100  # text tensors per time series
+
+
+def reduce_to_2d(arr):
+    """Given a np.npdarray with nDims > 2, reduce it to 2d.
+
+    It does this by selecting the zeroth coordinate for every dimension greater
+    than two.
+
+    Args:
+      arr: a numpy ndarray of dimension at least 2.
+
+    Returns:
+      A two-dimensional subarray from the input array.
+
+    Raises:
+      ValueError: If the argument is not a numpy ndarray, or the dimensionality
+        is too low.
+    """
+    if not isinstance(arr, np.ndarray):
+        raise ValueError("reduce_to_2d requires a numpy.ndarray")
+
+    ndims = len(arr.shape)
+    if ndims < 2:
+        raise ValueError("reduce_to_2d requires an array of dimensionality >=2")
+    # slice(None) is equivalent to `:`, so we take arr[0,0,...0,:,:]
+    slices = ([0] * (ndims - 2)) + [slice(None), slice(None)]
+    return arr[slices]
+
+
+def reduce_and_jsonify(text_ndarr):
+    """Take a numpy.ndarray containing strings, and convert it into a
+    json-compatible list, also squashing it to two dimensions if necessary.
+
+    If the ndarray contains a single scalar string, then that ndarray is
+    converted to a list. If it contains an array of strings,
+    that array is converted to a list.  If the array contains dimensionality
+    greater than 2, all but two of the dimensions are removed, and a squashed
+    boolean is set to true.  Returned is a list, the shape of the original
+    array, and a boolean indicating squashsing has occured.
+
+    Args:
+      text_arr: A numpy.ndarray containing strings.
+
+    Returns:
+        a tuple containing:
+            The JSON-compatible list
+            The shape of the array
```

Comment:
indicate which shape this is referring to. I think now it refers to the original ndarray shape.

Reply:
done